python/pylibraft/pylibraft/matrix/select_k.pyx

```diff
@@ -0,0 +1,133 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# cython: profile=False
+# distutils: language = c++
+# cython: embedsignature = True
+# cython: language_level = 3
+
+from cython.operator cimport dereference as deref
+from libc.stdint cimport int64_t
+from libcpp cimport bool
+
+import numpy as np
+
+from pylibraft.common import auto_convert_output, cai_wrapper, device_ndarray
+from pylibraft.common.handle import auto_sync_handle
+from pylibraft.common.input_validation import is_c_contiguous
+
+from pylibraft.common.cpp.mdspan cimport (
+    device_matrix_view,
+    host_matrix_view,
+    make_device_matrix_view,
+    make_host_matrix_view,
+    row_major,
+)
+from pylibraft.common.cpp.optional cimport optional
+from pylibraft.common.handle cimport device_resources
+from pylibraft.common.mdspan cimport get_dmv_float, get_dmv_int64
+from pylibraft.matrix.cpp.select_k cimport select_k as c_select_k
+
+
+@auto_sync_handle
+@auto_convert_output
+def select_k(dataset, k=None, distances=None, indices=None, select_min=True,
+             handle=None):
+    """
+    Selects the top k items from each row in a matrix
+
+
+    Parameters
+    ----------
+    dataset : array interface compliant matrix, row-major layout,
+        shape (n_rows, dim). Supported dtype [float]
+    k : int
+        Number of items to return for each row.  Optional if indices or
+        distances arrays are given (in which case their second dimension
+        is k).
+    distances :  Optional array interface compliant matrix shape
+                (n_rows, k), dtype float. If supplied,
+                distances will be written here in-place. (default None)
+    indices :  Optional array interface compliant matrix shape
+                (n_rows, k), dtype int64_t. If supplied, neighbor
+                indices will be written here in-place. (default None)
+    select_min: : bool
+        Whether to select the minimum or maximum K items
+
+    {handle_docstring}
+
+    Returns
+    -------
+    distances: array interface compliant object containing resulting distances
+               shape (n_rows, k)
+
+    indices: array interface compliant object containing resulting indices
+             shape (n_rows, k)
+
+    Examples
+    --------
+
+    >>> import cupy as cp
+
+    >>> from pylibraft.matrix import select_k
+
+    >>> n_features = 50
+    >>> n_rows = 1000
+
+    >>> queries = cp.random.random_sample((n_rows, n_features),
+    ...                                   dtype=cp.float32)
+    >>> k = 40
+    >>> distances, ids = select_k(queries, k)
+    >>> distances = cp.asarray(distances)
+    >>> ids = cp.asarray(ids)
+    """
+
+    dataset_cai = cai_wrapper(dataset)
```

Comment:
Wow- this whole function is quite elegant. For cython, anyways... but we've managed to isolate a lot of the boilerplate when working w/ mdspan/optional. It's great!

Reply:
I think so too!  I like how we're getting to the point where the cython code is just a thin wrapper over c++. With the CAI wrappers, the `get_dmv_float` code, the decorators to convert the output etc - I think its really getting quite nice here